packages/core/src/utils/oidc-provider-event-listener.ts

```diff
@@ -49,3 +49,27 @@ export const grantSuccessListener = async (
     scope,
   });
 };
+
+export const grantErrorListener = async (
+  ctx: KoaContextWithOIDC & WithLogContext & { body: GrantBody },
+  error: errors.OIDCProviderError
+) => {
+  const {
+    oidc: {
+      entities: { Client: client },
+      params,
+    },
+  } = ctx;
+  ctx.addLogContext({
+    applicationId: client?.clientId,
+  });
+
+  const grantType = params?.grant_type;
+  const type: LogType =
+    grantType === GrantType.AuthorizationCode ? 'CodeExchangeToken' : 'RefreshTokenExchangeToken';
```

Comment:
You can consider extracting this `getType` logic into a static method

Reply:
Refactored. PTAL.